teachertool/src/services/makecodeEditorService.ts

```diff
@@ -0,0 +1,62 @@
+interface PendingMessage {
+    original: pxt.editor.EditorMessageRequest;
+    handler: (response: any) => void;
+}
+
+let makecodeEditorRef: HTMLIFrameElement | undefined;
+const messageQueue: pxt.editor.EditorMessageRequest[] = [];
+let nextId: number = 0;
+let pendingMessages: {[index: string]: PendingMessage} = {};
+
+const onMessageReceived = (event: MessageEvent) => {
+    const data = event.data as pxt.editor.EditorMessageRequest;
+    if (data.type === "pxteditor" && data.id && pendingMessages[data.id]) {
+        const pending = pendingMessages[data.id];
+        pending.handler(data);
+        delete pendingMessages[data.id];
+        return;
+    }
+
+    console.log("Received message from iframe:", data);
+}
+
+const sendMessageAsync = (message?: any) => {
+    return new Promise(resolve => {
+        const sendMessageCore = (message: any) => {
+            message.response = true;
+            message.id = nextId++ + "";
+            pendingMessages[message.id] = {
+                original: message,
+                handler: resolve
+            };
+            makecodeEditorRef!.contentWindow!.postMessage(message, "*");
+        }
+
+        messageQueue.push(message);
+        if (makecodeEditorRef) {
+            while (messageQueue.length) {
+                sendMessageCore(messageQueue.shift());
+            }
+        }
+    });
+}
+
+export const setEditorRef = (ref: HTMLIFrameElement | undefined) => {
+    if (ref) {
+        makecodeEditorRef = ref;
+        window.addEventListener("message", onMessageReceived);
```

Comment:
I thought there was a comment on this already but maybe it's considered outdated now. Just to reiterate, we should clear the message queue here now that the ref is set (which, as far as this change is concerned, is indicating that the iframe is "ready")

Reply:
If following skillmap's convention, you can do that by calling sendMessageAsync with no parameter passed in, but you'll need to update your sendMessageAsync function to handle the message being undefined (just a check before adding it to the queue)